Address Point Match Check

savannahostrowski opened the pull request:
This check identifies Point objects in OSM that have a specified street number (addr:housenumber)
but no specified street name (addr:street). No specified street name refers to either having a null
value for the street name key, or no street name key present at all.

This pull request includes: all check logic, unit tests, and documentation.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/points/AddressPointMatchCheck.java:
```diff
@@ -0,0 +1,136 @@
+                // And if the street name key has a value of null
+                && (!object.getTag(POINT_STREET_NAME_KEY).isPresent()
+                        // Or if the street name key is not present
+                        || !object.getTags().containsKey(POINT_STREET_NAME_KEY));
```
The condition (`!object.getTag(POINT_STREET_NAME_KEY).isPresent() || !object.getTags().containsKey(POINT_STREET_NAME_KEY)`) checking for street name tag could be simplified by `Strings.isEmpty(object.tag(POINT_STREET_NAME_KEY))`.

`object.tag(POINT_STREET_NAME_KEY)` would return `null` if tag does not exist. If it exists, it's value would be returned. `Strings.isEmpty` would return true if given value is null or empty string.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/points/AddressPointMatchCheck.java:
```diff
@@ -0,0 +1,136 @@
+                .filter(interiorEdge -> interiorEdge.getTag(EDGE_STREET_NAME_KEY).isPresent())
+                .collect(Collectors.toSet());
+
+        final Set<String> streetNameList = new HashSet<>();
```
This doesn't have to be declared here.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/points/AddressPointMatchCheck.java:
```diff
@@ -0,0 +1,136 @@
+        {
+            // Add all interior Point street names to the list of candidate street names
+            points.forEach(interiorPoint -> streetNameList
+                    .add(interiorPoint.getTags().get(POINT_STREET_NAME_KEY)));
```
Street names could be collected by `points.forEach(point -> point.tag(POINT_STREET_NAME_KEY)).collect(Collectors.toSet())`.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/points/AddressPointMatchCheck.java:
```diff
@@ -0,0 +1,136 @@
+        {
+            // Add all interior Edge street names to the list of candidate street names
+            edges.forEach(interiorEdge -> streetNameList
+                    .add(interiorEdge.getTags().get(EDGE_STREET_NAME_KEY)));
```
Please see comment above about collecting street names.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/points/AddressPointMatchCheck.java:
```diff
@@ -0,0 +1,136 @@
+    private static final long serialVersionUID = 1L;
+    public static final String NO_STREET_NAME_POINT_INSTRUCTIONS = "This node, {0,number,#}, has "
+            + "no street name specified in the address. The street name should likely "
+            + "be one of {1}. These names were derived from nearby Points.";
```
OSM doesn't have points. I'd suggest referring them as nodes.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/points/AddressPointMatchCheck.java:
```diff
@@ -0,0 +1,136 @@
+            + "be one of {1}. These names were derived from nearby Points.";
+    public static final String NO_STREET_NAME_EDGE_INSTRUCTIONS = "This node, {0,number,#}, has "
+            + "no street name specified in the address. The street name should likely "
+            + "be one of {1}. These names were derived from nearby Edges.";
```
OSM doesn't have edges. I'd suggest referring them as ways.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/points/AddressPointMatchCheck.java:
```diff
@@ -0,0 +1,136 @@
+            + "no street name specified in the address. The street name should likely "
+            + "be one of {1}. These names were derived from nearby Edges.";
+    public static final String NO_SUGGESTED_NAMES_INSTRUCTIONS = "This node, {0,number,#}, has "
+            + "no street name specified in the address. No suggestions names were found.";
```
I'd suggest mentioning the tag key that is missing the value here.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/points/AddressPointMatchCheck.java:
```diff
@@ -0,0 +1,136 @@
+    private static final String ADDRESS_STREET_NUMBER_KEY = AddressHousenumberTag.KEY;
+    private static final String POINT_STREET_NAME_KEY = AddressStreetTag.KEY;
+    private static final String EDGE_STREET_NAME_KEY = NameTag.KEY;
+    private static final double BOUNDS_SIZE_DEFAULT = 150.0;
```
This seems a bit high. This would create a rectangle with ~300m long sides, which would cover a relatively big area. I'd suggest running with lower values and checking the results. I'd bet with a value like 25 here you'd be able to generate the same results.

isabelleh replied:
In some cases where an address belongs to a campus/hospital/large building it can be quite far from any nearby named roads so I think 25 would be a low value. Something closer to 75 might improve performance but catch more of the large building cases.

kasims44 replied:
I did a sensitivity analysis on different values and found that 75m yields the best results as a default value for the bounding box size. Any lower and the number of street recommendations drops significantly while increasing the size beyond 75 yields less and less significant results the larger the box becomes.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/points/AddressPointMatchCheck.java:
```diff
@@ -0,0 +1,136 @@
+        final Rectangle box = point.getLocation().boxAround(boundsSize);
+
+        // Get all Points inside the bounding box
+        final Iterable<Point> interiorPoints = point.getAtlas().pointsWithin(box);
```
Name `interiorPoints` is a bit misleading (geometries have interior points). I'd suggest renaming it to `nearbyPoints`.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/points/AddressPointMatchCheck.java:
```diff
@@ -0,0 +1,136 @@
+        final Iterable<Point> interiorPoints = point.getAtlas().pointsWithin(box);
+        // Convert the Iterable into a Stream for filtering, remove Points that have null as their
+        // street name as they cannot be candidates for a street for the Point of interest
+        final Set<Point> points = StreamSupport.stream(interiorPoints.spliterator(), false)
```
You can say `Iterable.stream(point.getAtlas().pointsWithin(box))` to stream points within an area.

savannahostrowski replied:
@mertemin Is there a nice way to collect an Iterables object into a Set to retain unique street names? Also, is there a concise way of checking whether Iterables are empty? I know Guava has this functionality but in straight Java, I don't think there's a pretty way of checking/doing those things.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/points/AddressPointMatchCheck.java:
```diff
@@ -0,0 +1,136 @@
+                .collect(Collectors.toSet());
+
+        // Get all Edges that intersect or are contained within the bounding box
+        final Iterable<Edge> interiorEdges = point.getAtlas().edgesIntersecting(box);
```
Same concern here. I'd suggest renaming it to `nearbyEdges`.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/points/AddressPointMatchCheck.java:
```diff
@@ -0,0 +1,136 @@
+        final Iterable<Edge> interiorEdges = point.getAtlas().edgesIntersecting(box);
+        // Convert the Iterable into a Stream for filtering, remove Edge that have null as their
+        // street name as they cannot be candidates for a street for the Point of interest
+        final Set<Edge> edges = StreamSupport.stream(interiorEdges.spliterator(), false)
```
In a similar sense, `Iterables.stream` could be utilized here.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/points/AddressPointMatchCheck.java:
```diff
@@ -0,0 +1,136 @@
+import org.openstreetmap.atlas.utilities.scalars.Distance;
+
+/**
+ * Tests for {@link AddressPointMatchCheck}
```
Needs an update.

mertemin commented on src/test/java/org/openstreetmap/atlas/checks/validation/points/AddressPointMatchCheckTestRule.java:
```diff
@@ -0,0 +1,98 @@
+ * @author savannahostrowski
+ */
+
+public class AddressPointMatchCheckTestRule extends CoreTestRule
```
I'd suggest writing tests with empty street name tag value for the initial point.

savannahostrowski commented:
Updates to add Associated Street Relation functionality (as mentioned by @isabelleh , and address @mertemin's comments).

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/points/AddressPointMatchCheck.java:
```diff
@@ -0,0 +1,166 @@
+            NO_SUGGESTED_NAMES_INSTRUCTIONS);
+    private static final String ADDRESS_STREET_NUMBER_KEY = AddressHousenumberTag.KEY;
+    private static final String POINT_STREET_NAME_KEY = AddressStreetTag.KEY;
+    private static final String EDGE_STREET_NAME_KEY = NameTag.KEY;
```
Please consider not redefining these static values. `*.KEY` format is already short and clean.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/points/AddressPointMatchCheck.java:
```diff
@@ -0,0 +1,166 @@
+    private static final String EDGE_STREET_NAME_KEY = NameTag.KEY;
+    private static final String STREET_RELATION_ROLE = "street";
+    private static final String ASSOCIATED_STREET_RELATION = "associatedStreet";
+    private static final double BOUNDS_SIZE_DEFAULT = 150.0;
```
There was a comment thread about this value and 75m was found to be the optimum value. This value is still 150m, which makes the size width/height of a bounding box ~300m.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/points/AddressPointMatchCheck.java:
```diff
@@ -0,0 +1,166 @@
+                && object.getTag(ADDRESS_STREET_NUMBER_KEY).isPresent()
+                // And if the street name key has a value of null or if the street name key is not
+                // present
+                && Strings.isEmpty(object.tag(POINT_STREET_NAME_KEY));
```
Please consider dropping the previous condition (`object.getTag(ADDRESS_STREET_NUMBER_KEY).isPresent()`) and replacing `Strings.isEmpty(object.tag(POINT_STREET_NAME_KEY))` with `Strings.isNullOrEmpty(object.tag(POINT_STREET_NAME_KEY))`

`Strings.isNullOrEmpty(object.tag(POINT_STREET_NAME_KEY))` will return
- true, if tag does not exist
- true, if tag exists, but null
- true, if tag exists, but empty
- false, if tag exists and has non-empty value

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/points/AddressPointMatchCheck.java:
```diff
@@ -0,0 +1,166 @@
+        // Where type=associatedStreet
+        final Set<Relation> relations = point.relations().stream().filter(
+                relation -> relation.tag(RelationTypeTag.KEY).equals(ASSOCIATED_STREET_RELATION))
+                .collect(Collectors.toSet());
```
Please consider replacing the logic of this method with something like:
```
point.relations().stream().filter(relation -> relation.tag(RelationTypeTag.KEY).equals(ASSOCIATED_STREET_RELATION))
.anyMatch(relation -> relation.members().stream().anyMatch(member -> member.getRole().equals(STREET_RELATION_ROLE) && member.getEntity().getType().equals(ItemType.EDGE)))
```

Such call would return `true` if point is a part of a relation where relation has an `Edge` member with `street` role.

Java's `Stream` has powerful methods like `anyMatch` that could be used to find if any item is matching the condition in a stream of items. Collecting items into sets here is not necessary.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/points/AddressPointMatchCheck.java:
```diff
@@ -0,0 +1,166 @@
+        // names
+        final Set<String> points = Iterables.stream(point.getAtlas().pointsWithin(box))
+                .filter(nearbyPoint -> nearbyPoint.getTag(POINT_STREET_NAME_KEY).isPresent())
+                .map(nearbyPoint -> nearbyPoint.tag(POINT_STREET_NAME_KEY)).collectToSet();
```
`nearbyPoint`'s street name tag is accessed twice here if it exists. I'd suggest mapping street name tags first (via `tag` method) and then filtering out nulls later and finally collecting them into a set.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/points/AddressPointMatchCheck.java:
```diff
@@ -0,0 +1,166 @@
+        // names
+        final Set<String> edges = Iterables.stream(point.getAtlas().edgesIntersecting(box))
+                .filter(nearbyEdge -> nearbyEdge.getTag(EDGE_STREET_NAME_KEY).isPresent())
+                .map(nearbyEdge -> nearbyEdge.tag(EDGE_STREET_NAME_KEY)).collectToSet();
```
Same concern here. Tag is looked up and accessed twice if it exists.

mertemin commented on src/main/java/org/openstreetmap/atlas/checks/validation/points/AddressPointMatchCheck.java:
```diff
@@ -0,0 +1,166 @@
+
+public class AddressPointMatchCheck extends BaseCheck
+{
+    private static final long serialVersionUID = 1L;
```
Please update with a generated UID.

savannahostrowski commented:
Refactoring to `hasAssociatedStreetRelation()`, and finding points and edges in bounding box to avoid a double call to get the tag on each object. Also updated documentation accordingly, and added a serial version UID.
